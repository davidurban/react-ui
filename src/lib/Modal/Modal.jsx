import PropTypes from 'prop-types';
import React from 'react';
import Button from '../Button';
import { withTranslationContext } from '../Translation/index';
import styles from './Modal.scss';

class Modal extends React.Component {
  constructor(props) {
    super(props);

    this.pressEscapeHandler = this.pressEscapeHandler.bind(this);
  }

  componentWillMount() {
    window.document.addEventListener('keydown', this.pressEscapeHandler, false);
  }

  componentWillUnmount() {
    window.document.removeEventListener('keydown', this.pressEscapeHandler, false);
  }

  pressEscapeHandler(e) {
    if (e.keyCode === 27) {
      this.props.closeHandler();
    }
  }

  render() {
    return (
      <div
        className={styles.overlay}
        onClick={this.props.closeHandler}
        role="presentation"
      >
        <div
          className={styles.root}
          onClick={(e) => {
            e.stopPropagation();
          }}
          role="presentation"
        >
          <div className={styles.head}>
            <h3 className={styles.headTitle}>
              {this.props.title}
            </h3>
          </div>
          <div className={styles.body}>
            {this.props.children}
          </div>
          <div className={styles.footer}>
            <span className={styles.button}>
              {this.props.actions.map(action => (
                <Button
                  clickHandler={action.clickHandler}
                  disabled={action.disabled}
                  key={action.label}
                  label={action.label}
                  loading={action.loading}
                  priority="primary"
                  variant={action.variant}
                />
              ))}
            </span>
            <span className={styles.button}>
              <Button
                clickHandler={this.props.closeHandler}
                label={this.props.translations.close}
              />
            </span>
          </div>
        </div>
      </div>
    );
  }
}

Modal.defaultProps = {
  actions: [],
};

Modal.propTypes = {
  actions: PropTypes.arrayOf(PropTypes.shape({
    clickHandler: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
    label: PropTypes.string.isRequired,
    loading: PropTypes.bool,
    variant: PropTypes.string,
  })),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  closeHandler: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  translations: PropTypes.shape({
    close: PropTypes.string.isRequired,
  }).isRequired,
};

export default withTranslationContext(Modal);
