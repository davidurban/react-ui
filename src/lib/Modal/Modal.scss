@import '../Theme/mixins';
@import '../Theme/variables';

.overlay {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 99;
  width: 100vw;
  height: 100vh;
  background-color: rgba(0, 0, 0, 0.5);
}

.root {
  position: fixed;
  top: 50%;
  left: 50%;
  z-index: 100;
  display: flex;
  flex-direction: column;
  width: 46.875rem;
  max-width: calc(100% - (2 * #{$rui-modal-offset-x}));
  max-height: calc(100% - (2 * #{$rui-modal-offset-x}));
  overflow-y: auto;
  border-radius: $rui-modal-border-radius;
  background-color: $white;
  box-shadow: 0 0 0.312rem 0 transparentize($black, 0.5);
  transform: translate(-50%, -50%);
}

.head {
  display: flex;
  flex: none;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  padding: $rui-modal-offset-y 0 $rui-modal-offset-y $rui-modal-offset-x;
}

.body {
  padding: $rui-modal-offset-y $rui-modal-offset-x;
  overflow-y: auto;
}

.headTitle {
  margin: 0;
}

.footer {
  display: flex;
  flex: none;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  padding: $rui-modal-offset-y $rui-modal-offset-x;
  border-bottom-right-radius: $rui-modal-border-radius;
  border-bottom-left-radius: $rui-modal-border-radius;
  background-color: $gray-light-midtone;
}

.button {
  margin: 0.25rem 0.75rem;
}

.isContentOverflowing .head {
  box-shadow: 0 0.4rem 1.7rem transparentize($black, 0.8);
}

.isContentOverflowing .footer {
  box-shadow: 0 -0.4rem 2.7rem transparentize($black, 0.8);
}
