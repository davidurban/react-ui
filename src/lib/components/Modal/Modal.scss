@import '../../shared/settings/colors.deprecated';
@import './variables';

.overlay {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 99;
  width: 100vw;
  height: 100vh;
  background-color: rgba(0, 0, 0, 0.5);
}

.root {
  position: fixed;
  top: 50%;
  left: 50%;
  z-index: 100;
  display: flex;
  flex-direction: column;
  width: 46.875rem;
  max-width: calc(100% - (2 * #{$modal-offset-x}));
  max-height: calc(100% - (2 * #{$modal-offset-x}));
  overflow-y: auto;
  border-radius: $modal-border-radius;
  background-color: $white;
  box-shadow: 0 0 0.312rem 0 transparentize($black, 0.5);
  transform: translate(-50%, -50%);
}

.head {
  display: flex;
  flex: none;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  padding: $modal-offset-y 0 $modal-offset-y $modal-offset-x;
}

.body {
  padding: $modal-offset-y $modal-offset-x;
  overflow-y: auto;
}

.headTitle {
  margin: 0;
}

.footer {
  display: flex;
  flex: none;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  padding: $modal-offset-y $modal-offset-x;
  border-bottom-right-radius: $modal-border-radius;
  border-bottom-left-radius: $modal-border-radius;
  background-color: $gray-light-midtone;
}

.button {
  margin: 0.25rem 0.75rem;
}

.isContentOverflowing .head {
  box-shadow: 0 0.4rem 1.7rem transparentize($black, 0.8);
}

.isContentOverflowing .footer {
  box-shadow: 0 -0.4rem 2.7rem transparentize($black, 0.8);
}
