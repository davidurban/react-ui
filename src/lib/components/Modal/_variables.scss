@import '../../shared/settings/borders';
@import '../../shared/settings/colors';
@import '../../shared/settings/offsets';
@import '../../shared/settings/typography';
@import '../../shared/settings/zindexes';

// 1. Intentionally using direct color to prevent custom properties from being passed into `rgba` function.

$modal-width: 46rem;
$modal-padding-x: map-get($offset-values, 5);
$modal-padding-y: map-get($offset-values, 3);
$modal-border-radius: $border-radius;
$modal-head-title-font-size: $typography-size-1;
$modal-head-box-shadow: 0 0.2rem 0.8rem rgba(0, 0, 0, 0.1); // 1.
$modal-footer-box-shadow: 0 -0.2rem 0.8rem rgba(0, 0, 0, 0.1); // 1.
$modal-zindex: $zindex-modal;
$modal-overlay-zindex: $zindex-modal-overlay;
