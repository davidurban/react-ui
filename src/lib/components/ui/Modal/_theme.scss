$modal-background: var(--rui-modal-background);
$modal-box-shadow: var(--rui-modal-box-shadow);
$modal-footer-background: var(--rui-modal-footer-background);
$modal-overlay-background: var(--rui-modal-overlay-background);
