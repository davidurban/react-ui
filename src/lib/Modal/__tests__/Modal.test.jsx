import React from 'react';
import {
  mount,
  shallow,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import Modal from '../index';

jest.mock('../../Icon/load-material-design-icons');

describe('rendering', () => {
  it('renders correctly with mandatory props only', () => {
    const tree = shallow((
      <Modal
        closeHandler={() => {}}
        title="Modal title"
      >
        Modal content
      </Modal>
    ));

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow((
      <Modal
        actions={[
          {
            clickHandler: () => {},
            label: 'Action',
          },
        ]}
        closeHandler={() => {}}
        title="Modal title"
      >
        Modal content
      </Modal>
    ));

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls closeHandler() on Close button click', () => {
    const spy = sinon.spy();
    const component = mount((
      <Modal
        closeHandler={spy}
        title="Modal title"
      >
        Modal content
      </Modal>
    ));

    component.find('Button').simulate('click');
    expect(spy.calledOnce).toEqual(true);
  });
});

