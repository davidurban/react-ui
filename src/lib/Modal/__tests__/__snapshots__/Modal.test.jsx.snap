// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with all props except translations 1`] = `
<WithTranslationContextComponent
  actions={
    Array [
      Object {
        "clickHandler": [Function],
        "label": "Action",
      },
    ]
  }
  closeHandler={[Function]}
  title="Modal title"
  translations={null}
>
  <Modal
    actions={
      Array [
        Object {
          "clickHandler": [Function],
          "label": "Action",
        },
      ]
    }
    closeHandler={[Function]}
    title="Modal title"
    translations={
      Object {
        "close": "Close",
      }
    }
  >
    <div
      className="overlay"
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="root"
        onClick={[Function]}
        role="presentation"
      >
        <div
          className="head"
        >
          <h3
            className="headTitle"
          >
            Modal title
          </h3>
        </div>
        <div
          className="body"
        >
          Modal content
        </div>
        <div
          className="footer"
        >
          <span
            className="button"
          >
            <Button
              block={false}
              clickHandler={[Function]}
              disabled={false}
              icon={null}
              iconPosition="before"
              isLabelVisible={true}
              key="Action"
              label="Action"
              loading={false}
              priority="primary"
              size="medium"
              type="button"
              variant="default"
            >
              <button
                className="root
                      priorityPrimary
                      sizeMedium"
                disabled={false}
                onClick={[Function]}
                title={null}
                type="button"
              >
                <span>
                  Action
                </span>
              </button>
            </Button>
          </span>
          <span
            className="button"
          >
            <Button
              block={false}
              clickHandler={[Function]}
              disabled={false}
              icon={null}
              iconPosition="before"
              isLabelVisible={true}
              label="Close"
              loading={false}
              priority="default"
              size="medium"
              type="button"
              variant="default"
            >
              <button
                className="root
                      priorityDefault
                      sizeMedium"
                disabled={false}
                onClick={[Function]}
                title={null}
                type="button"
              >
                <span>
                  Close
                </span>
              </button>
            </Button>
          </span>
        </div>
      </div>
    </div>
  </Modal>
</WithTranslationContextComponent>
`;

exports[`rendering renders correctly with mandatory props only 1`] = `
<WithTranslationContextComponent
  closeHandler={[Function]}
  title="Modal title"
  translations={null}
>
  <Modal
    actions={Array []}
    closeHandler={[Function]}
    title="Modal title"
    translations={
      Object {
        "close": "Close",
      }
    }
  >
    <div
      className="overlay"
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="root"
        onClick={[Function]}
        role="presentation"
      >
        <div
          className="head"
        >
          <h3
            className="headTitle"
          >
            Modal title
          </h3>
        </div>
        <div
          className="body"
        >
          Modal content
        </div>
        <div
          className="footer"
        >
          <span
            className="button"
          />
          <span
            className="button"
          >
            <Button
              block={false}
              clickHandler={[Function]}
              disabled={false}
              icon={null}
              iconPosition="before"
              isLabelVisible={true}
              label="Close"
              loading={false}
              priority="default"
              size="medium"
              type="button"
              variant="default"
            >
              <button
                className="root
                      priorityDefault
                      sizeMedium"
                disabled={false}
                onClick={[Function]}
                title={null}
                type="button"
              >
                <span>
                  Close
                </span>
              </button>
            </Button>
          </span>
        </div>
      </div>
    </div>
  </Modal>
</WithTranslationContextComponent>
`;

exports[`rendering renders correctly with translations 1`] = `
<WithTranslationContextComponent
  closeHandler={[Function]}
  title="Modal title"
  translations={
    Object {
      "close": "Zavřít",
    }
  }
>
  <Modal
    actions={Array []}
    closeHandler={[Function]}
    title="Modal title"
    translations={
      Object {
        "close": "Zavřít",
      }
    }
  >
    <div
      className="overlay"
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="root"
        onClick={[Function]}
        role="presentation"
      >
        <div
          className="head"
        >
          <h3
            className="headTitle"
          >
            Modal title
          </h3>
        </div>
        <div
          className="body"
        >
          Modal content
        </div>
        <div
          className="footer"
        >
          <span
            className="button"
          />
          <span
            className="button"
          >
            <Button
              block={false}
              clickHandler={[Function]}
              disabled={false}
              icon={null}
              iconPosition="before"
              isLabelVisible={true}
              label="Zavřít"
              loading={false}
              priority="default"
              size="medium"
              type="button"
              variant="default"
            >
              <button
                className="root
                      priorityDefault
                      sizeMedium"
                disabled={false}
                onClick={[Function]}
                title={null}
                type="button"
              >
                <span>
                  Zavřít
                </span>
              </button>
            </Button>
          </span>
        </div>
      </div>
    </div>
  </Modal>
</WithTranslationContextComponent>
`;
